### PR TITLE
pinyin: Keep initial and final in internal representation

### PR DESCRIPTION
**Pull request by JohnAZoidberg**
parsers: Add more finals

Standalone o final is used in a couple of characters.
ue needs to be included because characters.map includes jué a lot.

---
hanzi: Remove characters with strange pinyin

Since we can't parse them, we can't deal with them in a useful manner.
When the internal representation of the Pinyin module is changed to
internal/final/tone we can't represent these characters at all anymore,
because we have to parse them.

When the

Strangeness:

- n with tone marker
- Standalone o
- Strange standalone finals:
  - ng
  - n
  - m
  - wong

I don't think it would be useful to adapt the parsers to include these
because we'd parse a lot of strange sequences.
Not sure if they are "allowed" in the "official" Pinyin?

The standalone o is not super strange, people actually use the character 哦

TODO:
- [ ] Modify the converter to generate this properly.
- [x] Don't remove commonly used characters

---

hanzi: Remove characters with strange pinyin

The parser has this information. Don't lose it.

This helps when implementing conversion to Zhuyin, otherwise we have to
parse the internal representation again.

---

TODO:

- [x] Make sure all documentation is updated. I might have forgotten to change some `foo/1` to `foo/2`. Not sure if elixir has an automated way to check that?

### CODE REVIEW

File: lib/pinyin.ex
Comment (JohnAZoidberg) [inline lib/pinyin.ex]: This is what I meant here:

> This remark is not entirely clear to me:
>
>> There's only the case that somebody provides the pinyin and wants to output pinyin again, where it would add an additional parse step.

Before we didn't have to parse, we just put the string into `%Pinyin`. Now we have to parse, to split it.
This is only used during compile time when parsing the `.map` file.
I didn't see a noticeable increase in compile.
I'll definitely try to measure that though.

File: lib/pinyin.ex
Comment (JohnAZoidberg) [inline lib/pinyin.ex]: One side-effect of allowing "o" as a standalone final.
Maybe we want to add another test case to have at least one that returns an error.

File: lib/pinyin.ex
Comment (mathsaey) [inline lib/pinyin.ex]: Nitpick: I'd change the order here (`initial`, `final`), to make it a bit more pleasant to read.

File: lib/pinyin.ex
Comment (mathsaey) [inline lib/pinyin.ex]: Same here.

File: lib/pinyin.ex
Comment (mathsaey) [inline lib/pinyin.ex]: I am not sure if this `from_marked` still makes sense after the changes.
If a user already split the string in an initial and final part we could expect they also separated the tone, after which they can just use `create/3`.

File: lib/pinyin.ex
Comment (mathsaey) [reply lib/pinyin.ex]: The main point of the `from_marked` and `from_numbered` functions was that they could be used to "skip" the parser if you already knew the type of Pinyin you were dealing with (e.g. you could use `from_marked` to parse a string like "hǎo3") . I think we should consider dropping these functions if they don't work together with the new representation.

Down the line, I could maybe introduce an option for the pinyin parser which only considers numbered/marked characters.

File: lib/pinyin.ex
Comment (mathsaey) [reply lib/pinyin.ex]: I would certainly add an extra test here, to ensure the tests cover this case.

File: lib/pinyin.ex
Comment (mathsaey) [inline lib/pinyin.ex]: Same remark, keep these in reading order.

File: lib/pinyin.ex
Comment (mathsaey) [inline lib/pinyin.ex]: And again :).

File: lib/pinyin/parsers.ex
Comment (mathsaey) [inline lib/pinyin/parsers.ex]: A bit nitpicky, but could you put this between "wo" and "wa" (and adjust the spacing) to be consistent with the `combination_finals` table?

File: lib/pinyin/parsers.ex
Comment (mathsaey) [inline lib/pinyin/parsers.ex]: Perhaps we can put this on a line below ve, to keep the table consistent.

**Review by mathsaey — CHANGES_REQUESTED** [review]
Looks good overall, although I have a few nitpicks. My only major remarks concerns the `from_marked` and `from_numbered` functions. If these functions use the parser under the hood, there is no major reason to use these over `read`. It might be better to just remove them and introduce a `:numbered` or `:marked` option to `read` further down the line.

**Comment by mathsaey** [comment]
> Make sure all documentation is updated. I might have forgotten to change some foo/1 to foo/2. Not sure if elixir has an automated way to check that?

`mix docs` should provide a warning if a link is broken.

File: lib/pinyin.ex
Comment (JohnAZoidberg) [reply lib/pinyin.ex]: I initially did that but noticed in the iex it would show up in alphabetic order. (Yes, tone wouldn't be in the correct order either..)
Since the test doesn't seem to care about this and it's documentation, I can definitely make the order more readable.

**Comment by mathsaey** [comment]
I'm not sure what to do about the strange standalone finals you encountered. Ideally I'd prefer to find a way to parse these over throwing away information that was present in the unicode database.

File: lib/pinyin.ex
Comment (JohnAZoidberg) [reply lib/pinyin.ex]: > The main point of the `from_marked` and `from_numbered` functions was that they could be used to "skip" the parser if you already knew the type of Pinyin you were dealing with (e.g. you could use `from_marked` to parse a string like "hǎo3") . I think we should consider dropping these functions if they don't work together with the new representation.

Makes sense, let me see if that works out.

> Down the line, I could maybe introduce an option for the pinyin parser which only considers numbered/marked characters.

Hm, makes sense, too. You mean just a faster version that avoids parsing but is only able to convert between numbered and marked (and that only if the caller guarantees it's a valid pinyin syllable)?

**Comment by JohnAZoidberg** [comment]
> I'm not sure what to do about the strange standalone finals you encountered. Ideally I'd prefer to find a way to parse these over throwing away information that was present in the unicode database.

I agree. I had never seen them used, so my first instinct was that there's something wrong :D (I guess that's a bit too lazy)

Wikipedia says:

> Technically, i, u, ü without a following vowel are finals, not medials, and therefore take the tone marks, but they are more concisely displayed as above. In addition, ê [ɛ] (欸; 誒) and syllabic nasals m (呒, 呣), n (嗯, 唔), ng (嗯, 𠮾) are used as [interjections](https://en.wikipedia.org/wiki/Interjection).

So the m, n, ng are actually valid, as is the ê.

File: lib/pinyin.ex
Comment (mathsaey) [reply lib/pinyin.ex]: > > Down the line, I could maybe introduce an option for the pinyin parser which only considers numbered/marked characters.
>
> Hm, makes sense, too. You mean just a faster version that avoids parsing but is only able to convert between numbered and marked (and that only if the caller guarantees it's a valid pinyin syllable)?

It's not about the speed but about the guarantees the function gives you. `from_marked` and `from_numbered` were very simple, so they could handle weird cases not handled by the parser. For instance, `from_marked` would correctly parse "hao3" as hao + the number 3 while `read` would always parse it as "hao" with the third tone.

While I don't think it is very common to mix tone marked pinyin with numbers without proper spaces I think it's might be inconvenient to lose that functionality. Since the parser already has a different code path for numbered and marked pinyin I might introduce a way to use the parser and specify whether the pinyin you provide is marked or numbered.

For now, I would propose to remove `from_marked` and `from_numbered` and I'll put the parser adjustements on my TODO list. I'll see if I can ever get to it 😅.

**Comment by mathsaey** [comment]
> > I'm not sure what to do about the strange standalone finals you encountered. Ideally I'd prefer to find a way to parse these over throwing away information that was present in the unicode database.
>
> I agree. I had never seen them used, so my first instinct was that there's something wrong :D (I guess that's a bit too lazy)

It's a solution, but it will probably be a pain to maintain the `convert_unihan_reading.exs` script. Especially if we add other dictionaries to the character map down the line.

> So the m, n, ng are actually valid, as is the ê.

Classic 😅. I would propose to add these to `standalone_finals` table. The parser only tries these after it failed to create a "standard" intial + final word, so it should only cause issues when parsing normal text mixed with pinyin, but that can be solved by using the parser in `:words` mode.

**Comment by JohnAZoidberg** [comment]
> > Make sure all documentation is updated. I might have forgotten to change some foo/1 to foo/2. Not sure if elixir has an automated way to check that?
> 
> `mix docs` should provide a warning if a link is broken.

No warning when I intentionally break something:

image

image

File: lib/pinyin/parsers.ex
Comment (JohnAZoidberg) [reply lib/pinyin/parsers.ex]: oh okay I get it. I didn't see the pattern here.

File: lib/pinyin.ex
Comment (JohnAZoidberg) [reply lib/pinyin.ex]: Done.

File: lib/pinyin.ex
Comment (JohnAZoidberg) [reply lib/pinyin.ex]: Done.

File: lib/pinyin/parsers.ex
Comment (JohnAZoidberg) [reply lib/pinyin/parsers.ex]: Done.

File: lib/pinyin/parsers.ex
Comment (JohnAZoidberg) [reply lib/pinyin/parsers.ex]: Done

**Comment by JohnAZoidberg** [comment]
> > So the m, n, ng are actually valid, as is the ê.
> 
> Classic sweat_smile. I would propose to add these to `standalone_finals` table. The parser only tries these after it failed to create a "standard" intial + final word, so it should only cause issues when parsing normal text mixed with pinyin, but that can be solved by using the parser in `:words` mode.

If I do that it parses `foo na foo` as `n` and `a`... Which is not what we want.
There's no way to fix this, as far as I can tell, if we still want to be able to parse `foo nihao foo`. (same with `ma`).

Maybe the best solution is to avoid parsing them when parsing pinyin text but be able to turn their hanzi into the correct pinyin. This is already the current behavior (without this PR), if I'm not mistaken.

**Comment by JohnAZoidberg** [comment]
Ok, I can parse everything now, except for the two that cause issues `n`, `m` and the accented versions of `ng`.
I think accented `ng` can be implemented. The others we'll just have to ignore, as I mentioned above.

File: lib/pinyin.ex
Comment (JohnAZoidberg) [inline lib/pinyin.ex]: That means adding it to `select_max/2`

Comment (JohnAZoidberg) [reply lib/pinyin.ex]: > That means adding it to `select_max/2`

Yeah that doesn't work...

Need a special case here.

**Comment by JohnAZoidberg** [comment]
> I think accented `ng` can be implemented.

Done

**Comment by mathsaey** [comment]
> > > Make sure all documentation is updated. I might have forgotten to change some foo/1 to foo/2. Not sure if elixir has an automated way to check that?
> > 
> > 
> > `mix docs` should provide a warning if a link is broken.
> 
> No warning when I intentionally break something:

Strange, I could have sworn I have seen some warnings like that, but I can't get them to work with an updated ex_doc either.

> Ok, I can parse everything now, except for the two that cause issues n, m and the accented versions of ng.
I think accented ng can be implemented. The others we'll just have to ignore, as I mentioned above.

I will try to find some time soon (tm) to properly look into this. Ideally, I can adjust the parser to handle these special cases. 

Thank you for all the work on this so far! I want to find some time to help you wrap up this PR so we can move onto merging zhuyin support :).

**Comment by JohnAZoidberg** [comment]
> Ideally, I can adjust the parser to handle these special cases.

It very much seems that the remaining sounds `m`, `n` are just exclamations. Which means they are never a syllable part of a word.
They are always a standalone syllable. We should be able to adjust the parser for that.

File: lib/hanzi/characters.map
Comment (JohnAZoidberg) [inline lib/hanzi/characters.map]: Oh no, this m isn't a single unicode code point like all the other accented "vowels".
It's U+006D and U+0304

File: lib/hanzi/characters.map
Comment (JohnAZoidberg) [inline lib/hanzi/characters.map]: This one is a single code-point U+1E3F

File: lib/hanzi/characters.map
Comment (JohnAZoidberg) [inline lib/hanzi/characters.map]: First m is U+1E3F, second one is U+006D and U+0300

Comment (JohnAZoidberg) [reply lib/hanzi/characters.map]: Meaning/Usage: https://en.wiktionary.org/wiki/%E5%91%A3

File: lib/hanzi/characters.map
Comment (JohnAZoidberg) [reply lib/hanzi/characters.map]: https://en.wiktionary.org/wiki/%E5%94%94

File: lib/hanzi/characters.map
Comment (JohnAZoidberg) [reply lib/hanzi/characters.map]: https://en.wiktionary.org/wiki/%E5%98%B8

**Comment by mathsaey** [comment]
I am working on the pinyin module and parser to handle the special cases you brought up, but it's fairly tricky, as you found out, so it will take me some time. Once I'm done, I'll push it to your PR so you can also give it a look before we merge this.

File: lib/hanzi/characters.map
Comment (mathsaey) [reply lib/hanzi/characters.map]: I'm working on the new version of `from_marked` which should be able to handle this, so we can parse this situation. I should have a few commits ready soon.

File: lib/pinyin.ex
Comment (mathsaey) [reply lib/pinyin.ex]: I've adjusted `marked` to handle this.

File: lib/pinyin.ex
Comment (mathsaey) [reply lib/pinyin.ex]: The new parser can parse numbered pinyin, marked pinyin, or both at the same time. Just need to make a nice frontend for it and I can add it to this PR.

File: lib/pinyin.ex
Comment (mathsaey) [reply lib/pinyin.ex]: I will remove these versions of `from_marked` and `from_numbered`. I'm just not sure what to do with the one argument version. They will either dissapear or become part of the frontend for the new parser.

**Comment by mathsaey** [comment]
All right, I managed to build a parser which can handle all the special cases you brought up (i.e. it can handle every entry in the unmodified `characters.map`). I just need to update the pinyin module to handle the new parsers, add some tests and change documentation and I'll add it to this PR so you can take a look before we merge this :).

**Comment by mathsaey** [comment]
@JohnAZoidberg I made quite a few changes to your original code. However, the new parser can now handle all the entries in `character.map`. Could you just give this a look and see if it's fine for your purposes?

File: lib/pinyin.ex
Comment (mathsaey) [reply lib/pinyin.ex]: I've made the "old" `from_marked` and `from_numbered` part of the new parser. In the updated versions the `from_` functions create a pinyin struct while the `read` functions create a pinyin list.

**Review by mathsaey — APPROVED** [review]

File: lib/pinyin/parsers.ex
Comment (JohnAZoidberg) [inline lib/pinyin/parsers.ex:70]: Curious in what sense this is better?

File: lib/pinyin.ex
Comment (JohnAZoidberg) [inline lib/pinyin.ex:37]: Very cool!

**Review by JohnAZoidberg — COMMENTED** [review]
Looks great! Good cleanup, too

File: lib/pinyin/parsers.ex
Comment (mathsaey) [reply lib/pinyin/parsers.ex:70]: It just avoids a string conversion and a parsing step by converting the codepoint straight to it's integer equivalent. It's not really related to the PR, but cleaned it up while I was at it.

**Pull request merged**
